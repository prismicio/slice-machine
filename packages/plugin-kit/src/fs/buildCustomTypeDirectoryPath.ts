import * as path from "node:path";

import {
	buildCustomTypeLibraryDirectoryPath,
	BuildCustomTypeLibraryDirectoryPathArgs,
} from "./buildCustomTypeLibraryDirectoryPath";

export type BuildCustomTypeDirectoryPathArgs = {
	customTypeID: string;
} & BuildCustomTypeLibraryDirectoryPathArgs;

export const buildCustomTypeDirectoryPath = (
	args: BuildCustomTypeDirectoryPathArgs,
): string => {
	return path.join(
		buildCustomTypeLibraryDirectoryPath(args),
		args.customTypeID,
	);
};
