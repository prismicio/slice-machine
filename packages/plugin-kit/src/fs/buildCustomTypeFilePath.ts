import * as path from "node:path";

import {
	buildCustomTypeDirectoryPath,
	BuildCustomTypeDirectoryPathArgs,
} from "./buildCustomTypeDirectoryPath";

export type BuildCustomTypeFilePathArgs = {
	filename: string;
} & BuildCustomTypeDirectoryPathArgs;

export const buildCustomTypeFilePath = (
	args: BuildCustomTypeFilePathArgs,
): string => {
	return path.join(buildCustomTypeDirectoryPath(args), args.filename);
};
