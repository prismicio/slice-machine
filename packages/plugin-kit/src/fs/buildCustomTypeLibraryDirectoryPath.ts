import { SliceMachineHelpers } from "../createSliceMachineHelpers";

import { CUSTOM_TYPES_LIBRARY_DIRNAME } from "./constants";

export type BuildCustomTypeLibraryDirectoryPathArgs = {
	absolute?: boolean;
	helpers: SliceMachineHelpers;
};

export const buildCustomTypeLibraryDirectoryPath = (
	args: BuildCustomTypeLibraryDirectoryPathArgs,
): string => {
	return args.absolute
		? args.helpers.joinPathFromRoot(CUSTOM_TYPES_LIBRARY_DIRNAME)
		: CUSTOM_TYPES_LIBRARY_DIRNAME;
};
