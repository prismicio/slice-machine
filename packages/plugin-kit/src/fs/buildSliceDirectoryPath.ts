import * as path from "node:path";

import { pascalCase } from "./lib/pascalCase";
import {
	resolveSliceModel,
	ResolveSliceModelArgs,
} from "./lib/resolveSliceModel";

import {
	buildSliceLibraryDirectoryPath,
	BuildSliceLibraryDirectoryPathArgs,
} from "./buildSliceLibraryDirectoryPath";

export type BuildSliceDirectoryPathArgs = BuildSliceLibraryDirectoryPathArgs &
	ResolveSliceModelArgs;

export const buildSliceDirectoryPath = async (
	args: BuildSliceDirectoryPathArgs,
): Promise<string> => {
	const model = await resolveSliceModel(args);

	return path.join(
		buildSliceLibraryDirectoryPath(args),
		pascalCase(model.name),
	);
};
