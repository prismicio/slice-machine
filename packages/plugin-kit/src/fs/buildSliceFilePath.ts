import * as path from "node:path";

import {
	buildSliceDirectoryPath,
	BuildSliceDirectoryPathArgs,
} from "./buildSliceDirectoryPath";

export type BuildSliceFilePathArgs = {
	filename: string;
} & BuildSliceDirectoryPathArgs;

export const buildSliceFilePath = async (
	args: BuildSliceFilePathArgs,
): Promise<string> => {
	return path.join(await buildSliceDirectoryPath(args), args.filename);
};
