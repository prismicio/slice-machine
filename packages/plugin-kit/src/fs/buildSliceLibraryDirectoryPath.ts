import * as path from "node:path";

import { SliceMachineHelpers } from "../createSliceMachineHelpers";

export type BuildSliceLibraryDirectoryPathArgs = {
	libraryID: string;
	absolute?: boolean;
	helpers: SliceMachineHelpers;
};

export const buildSliceLibraryDirectoryPath = (
	args: BuildSliceLibraryDirectoryPathArgs,
): string => {
	return args.absolute
		? args.helpers.joinPathFromRoot(args.libraryID)
		: path.join(args.libraryID);
};
