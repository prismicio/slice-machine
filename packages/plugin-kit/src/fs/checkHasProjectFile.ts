import { checkPathExists } from "./lib/checkPathExists";

import { SliceMachineHelpers } from "../createSliceMachineHelpers";

export type CheckHasProjectFileArgs = {
	filename: string;
	helpers: SliceMachineHelpers;
};

export const checkHasProjectFile = async (
	args: CheckHasProjectFileArgs,
): Promise<boolean> => {
	const filePath = args.helpers.joinPathFromRoot(args.filename);

	return checkPathExists(filePath);
};
