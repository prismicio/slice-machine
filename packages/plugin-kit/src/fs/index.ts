///////////////////////////////////////////////////////////////////////////////
//
// Project helpers
//
///////////////////////////////////////////////////////////////////////////////

export { checkIsTypeScriptProject } from "./checkIsTypeScriptProject";

export { upsertGlobalTypeScriptTypes } from "./upsertGlobalTypeScriptTypes";
export type { UpsertGlobalTypeScriptTypesArgs } from "./upsertGlobalTypeScriptTypes";

export { checkHasProjectFile } from "./checkHasProjectFile";
export type { CheckHasProjectFileArgs } from "./checkHasProjectFile";

export { writeProjectFile } from "./writeProjectFile";
export type { WriteProjectFileArgs } from "./writeProjectFile";

export { readProjectFile } from "./readProjectFile";
export type { ReadProjectFileArgs } from "./readProjectFile";

export { deleteProjectFile } from "./deleteProjectFile";
export type { DeleteProjectFileArgs } from "./deleteProjectFile";

export { readProjectEnvironment } from "./readProjectEnvironment";
export type {
	ReadProjectEnvironmentArgs,
	ReadEnvironmentReturnType,
} from "./readProjectEnvironment";

export { writeProjectEnvironment } from "./writeProjectEnvironment";
export type { WriteProjectEnvironmentArgs } from "./writeProjectEnvironment";

///////////////////////////////////////////////////////////////////////////////
//
// Slice helpers
//
///////////////////////////////////////////////////////////////////////////////

export { buildSliceDirectoryPath } from "./buildSliceDirectoryPath";
export type { BuildSliceDirectoryPathArgs } from "./buildSliceDirectoryPath";

export { buildSliceFilePath } from "./buildSliceFilePath";
export type { BuildSliceFilePathArgs } from "./buildSliceFilePath";

export { buildSliceLibraryDirectoryPath } from "./buildSliceLibraryDirectoryPath";
export type { BuildSliceLibraryDirectoryPathArgs } from "./buildSliceLibraryDirectoryPath";

export { readSliceLibrary } from "./readSliceLibrary";
export type {
	ReadSliceLibraryArgs,
	ReadSliceLibraryReturnType,
} from "./readSliceLibrary";

export { writeSliceFile } from "./writeSliceFile";
export type { WriteSliceFileArgs } from "./writeSliceFile";

export { writeSliceModel } from "./writeSliceModel";
export type { WriteSliceModelArgs } from "./writeSliceModel";

export { readSliceFile } from "./readSliceFile";
export type { ReadSliceFileArgs } from "./readSliceFile";

export { readSliceModel } from "./readSliceModel";
export type {
	ReadSliceModelArgs,
	ReadSliceModelReturnType,
} from "./readSliceModel";

export { renameSlice } from "./renameSlice";
export type { RenameSliceArgs } from "./renameSlice";

export { deleteSliceFile } from "./deleteSliceFile";
export type { DeleteSliceFileArgs } from "./deleteSliceFile";

export { deleteSliceDirectory } from "./deleteSliceDirectory";
export type { DeleteSliceDirectoryArgs } from "./deleteSliceDirectory";

///////////////////////////////////////////////////////////////////////////////
//
// Custom type helpers
//
///////////////////////////////////////////////////////////////////////////////

export { buildCustomTypeDirectoryPath } from "./buildCustomTypeDirectoryPath";
export type { BuildCustomTypeDirectoryPathArgs } from "./buildCustomTypeDirectoryPath";

export { buildCustomTypeFilePath } from "./buildCustomTypeFilePath";
export type { BuildCustomTypeFilePathArgs } from "./buildCustomTypeFilePath";

export { buildCustomTypeLibraryDirectoryPath } from "./buildCustomTypeLibraryDirectoryPath";
export type { BuildCustomTypeLibraryDirectoryPathArgs } from "./buildCustomTypeLibraryDirectoryPath";

export { readCustomTypeLibrary } from "./readCustomTypeLibrary";
export type {
	ReadCustomTypeLibraryArgs,
	ReadCustomTypeLibraryReturnType,
} from "./readCustomTypeLibrary";

export { writeCustomTypeFile } from "./writeCustomTypeFile";
export type { WriteCustomTypeFileArgs } from "./writeCustomTypeFile";

export { writeCustomTypeModel } from "./writeCustomTypeModel";
export type { WriteCustomTypeModelArgs } from "./writeCustomTypeModel";

export { readCustomTypeFile } from "./readCustomTypeFile";
export type { ReadCustomTypeFileArgs } from "./readCustomTypeFile";

export { readCustomTypeModel } from "./readCustomTypeModel";
export type {
	ReadCustomTypeModelArgs,
	ReadCustomTypeModelReturnType,
} from "./readCustomTypeModel";

export { renameCustomType } from "./renameCustomType";
export type { RenameCustomTypeArgs } from "./renameCustomType";

export { deleteCustomTypeFile } from "./deleteCustomTypeFile";
export type { DeleteCustomTypeFileArgs } from "./deleteCustomTypeFile";

export { deleteCustomTypeDirectory } from "./deleteCustomTypeDirectory";
export type { DeleteCustomTypeDirectoryArgs } from "./deleteCustomTypeDirectory";

export { readSliceTemplateLibrary } from "./readSliceTemplateLibrary";
export type { ReadSliceTemplateLibraryArgs } from "./readSliceTemplateLibrary";
