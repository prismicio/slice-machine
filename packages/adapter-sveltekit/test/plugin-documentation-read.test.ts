import { describe, expect, it } from "vitest";
import { createMockFactory } from "@prismicio/mock";
import * as fs from "node:fs/promises";
import * as path from "node:path";

/**
 * !!! DO NOT use this mock factory in tests !!!
 *
 * @remarks
 * Its seed is not specific to be used outside the most general cases.
 */
const mock = createMockFactory({ seed: import.meta.url });

// Slice model to be used in general tests.
const repeatableModel = mock.model.customType({
	id: "repeatable",
	repeatable: true,
});
const singleModel = mock.model.customType({
	id: "single",
	repeatable: false,
});

describe("PageSnippet", () => {
	describe("repeatable model", () => {
		it("return a snippet with TypeScript for TypeScript projects", async (ctx) => {
			await fs.writeFile(
				path.join(ctx.project.root, "tsconfig.json"),
				JSON.stringify({}),
			);

			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model: repeatableModel,
				},
			});

			const item = res.data.flat().find((item) => item.label === "Default");

			expect(item).toMatchSnapshot();
		});

		it("return a snippet with JavaScript", async (ctx) => {
			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model: repeatableModel,
				},
			});

			const item = res.data.flat().find((item) => item.label === "Default");

			expect(item).toMatchSnapshot();
		});
	});

	describe("single model", () => {
		it("return a snippet with TypeScript for TypeScript projects", async (ctx) => {
			await fs.writeFile(
				path.join(ctx.project.root, "tsconfig.json"),
				JSON.stringify({}),
			);

			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model: singleModel,
				},
			});

			const item = res.data.flat().find((item) => item.label === "Default");

			expect(item).toMatchSnapshot();
		});

		it("return a snippet with JavaScript", async (ctx) => {
			const res = await ctx.pluginRunner.callHook("documentation:read", {
				kind: "PageSnippet",
				data: {
					model: singleModel,
				},
			});

			const item = res.data.flat().find((item) => item.label === "Default");

			expect(item).toMatchSnapshot();
		});
	});
});
