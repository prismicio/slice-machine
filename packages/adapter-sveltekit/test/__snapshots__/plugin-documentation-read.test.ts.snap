// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PageSnippet > repeatable model > return a snippet with JavaScript 1`] = `
{
  "content": "## Create your Repeatable's page data fetcher

Add a new route by creating a \`src/routes/[uid]/+page.server.js\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/[uid]/+page.server.js\`.)

Paste in this code:

~~~js [src/routes/[uid]/+page.server.js]
import { createClient } from \\"$lib/prismicio\\";

export async function load({ params, fetch }) {
  const client = createClient({ fetch });

  const page = await client.getByUID(\\"repeatable\\", params.uid);

  return {
    page,
  };
}

export async function entries() {
  const client = createClient();

  const pages = await client.getAllByType(\\"repeatable\\");

  return pages.map((page) => {
    return { uid: page.uid };
  });
}
~~~

## Create your Repeatable's page component

In the route's directory, create a \`src/routes/[uid]/+page.svelte\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/[uid]/+page.svelte\`.)

Paste in this code:

~~~svelte [src/routes/[uid]/+page.svelte]
<script>
  import { SliceZone } from \\"@prismicio/svelte\\";

  import { components } from \\"$lib/slices\\";

  export let data;
</script>

<SliceZone slices={data.page.data.slices} {components} />
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/svelte-install) for more information.",
  "label": "Default",
}
`;

exports[`PageSnippet > repeatable model > return a snippet with TypeScript for TypeScript projects 1`] = `
{
  "content": "## Create your Repeatable's page data fetcher

Add a new route by creating a \`src/routes/[uid]/+page.server.ts\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/[uid]/+page.server.ts\`.)

Paste in this code:

~~~ts [src/routes/[uid]/+page.server.ts]
import { createClient } from \\"$lib/prismicio\\";

export async function load({ params, fetch }) {
  const client = createClient({ fetch });

  const page = await client.getByUID(\\"repeatable\\", params.uid);

  return {
    page,
  };
}

export async function entries() {
  const client = createClient();

  const pages = await client.getAllByType(\\"repeatable\\");

  return pages.map((page) => {
    return { uid: page.uid };
  });
}
~~~

## Create your Repeatable's page component

In the route's directory, create a \`src/routes/[uid]/+page.svelte\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/[uid]/+page.svelte\`.)

Paste in this code:

~~~svelte [src/routes/[uid]/+page.svelte]
<script>
  import { SliceZone } from \\"@prismicio/svelte\\";

  import { components } from \\"$lib/slices\\";

  export let data;
</script>

<SliceZone slices={data.page.data.slices} {components} />
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/svelte-install) for more information.",
  "label": "Default",
}
`;

exports[`PageSnippet > single model > return a snippet with JavaScript 1`] = `
{
  "content": "## Create your Single's page data fetcher

Add a new route by creating a \`src/routes/single/+page.server.js\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/single/+page.server.js\`.)

Paste in this code:

~~~js [src/routes/single/+page.server.js]
import { createClient } from \\"$lib/prismicio\\";

export async function load({ params, fetch }) {
  const client = createClient({ fetch });

  const page = await client.getSingle(\\"single\\");

  return {
    page,
  };
}

export async function entries() {
  return [{}];
}
~~~

## Create your Single's page component

In the route's directory, create a \`src/routes/single/+page.svelte\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/single/+page.svelte\`.)

Paste in this code:

~~~svelte [src/routes/single/+page.svelte]
<script>
  import { SliceZone } from \\"@prismicio/svelte\\";

  import { components } from \\"$lib/slices\\";

  export let data;
</script>

<SliceZone slices={data.page.data.slices} {components} />
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/svelte-install) for more information.",
  "label": "Default",
}
`;

exports[`PageSnippet > single model > return a snippet with TypeScript for TypeScript projects 1`] = `
{
  "content": "## Create your Single's page data fetcher

Add a new route by creating a \`src/routes/single/+page.server.ts\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/single/+page.server.ts\`.)

Paste in this code:

~~~ts [src/routes/single/+page.server.ts]
import { createClient } from \\"$lib/prismicio\\";

export async function load({ params, fetch }) {
  const client = createClient({ fetch });

  const page = await client.getSingle(\\"single\\");

  return {
    page,
  };
}

export async function entries() {
  return [{}];
}
~~~

## Create your Single's page component

In the route's directory, create a \`src/routes/single/+page.svelte\` file. (If the route should be nested in a child directory, you can create the file in a directory, like \`src/routes/marketing/single/+page.svelte\`.)

Paste in this code:

~~~svelte [src/routes/single/+page.svelte]
<script>
  import { SliceZone } from \\"@prismicio/svelte\\";

  import { components } from \\"$lib/slices\\";

  export let data;
</script>

<SliceZone slices={data.page.data.slices} {components} />
~~~

Make sure all of your import paths are correct. See the [install guide](https://prismic.io/docs/svelte-install) for more information.",
  "label": "Default",
}
`;
